### ISSUE TITLE
[TS] Context type is inconsistent

### ISSUE THREAD

**Issue by zaboco** [lukeed/uvu#86]
## Issue
The way the Context is typed right now is not really consistent with its real value. More specifically, in the test callback, it is assumed that the fields of the context are present, even if they aren't.

So, for example, the following code should not compile. But right now it does, and only fails at runtime with "Cannot read property 'name' of undefined".

```ts
interface Context {
  user: { name: string };
}

const mySuite = suite<Context>('my suite');

mySuite('fails with context', (context) => {
  assert.equal(context.user.name, 'admin');
});
```

## Possible solution
The suite function can have two definitions:
```ts
export function suite<T = Context>(title: string, context: T): uvu.Test<T>;
export function suite<T = Context>(title?: string, context?: T): uvu.Test<Partial<T>>;
```

So, if a default context is given, the full Context is given to the Test. If not, a `Partial<Context>` is given, so there will be compile errors when trying to access a context field.

Of course, this still has the shortcoming that it doesn't cover this case:
```ts
const mySuite = suite<{ user: { name: string } }>('my suite');

mySuite.before(context => {
  context.user = { name: 'admin' }
})

mySuite('works, but would not compile', (context) => {
  assert.equal(context.user.name, 'admin');
});
```

Not sure if it's even possible to infer that `context.user` exists. But I guess that this case can be replaced with passing the context in the suite constructor.

PS. I can make a PR if you think this helps.

**Comment by lukeed** [comment]
Hey, thanks, but this is _your_ `Context` being inaccurate.

Your interface defines a `user:` property instead of a `user?:` property. On uvu's end, whatever is defined is simply echoed, since _youre_ in control of the context itself.

If you scroll to the bottom of [this docs section](https://github.com/lukeed/uvu/blob/master/docs/api.uvu.md#context-1) you'll find a TS example that illustrates this, hopefully.

If not please circle back and let me know :D

**Comment by zaboco** [comment]
> Hey, thanks, but this is your Context being inaccurate.

Yes, of course, I could make the Context `{ user?: { name: string} }`, and it would be fine. But that's not the point.

The point is that _there are_ cases when the types don't represent the reality. One example is the above. In that case the actual behaviour is not the same as the expected behaviour. And that is the definition of a bug, in this case a bug at the type level. So, for the example above:
- **Expected behaviour**: The code does not compile
- **Actual behaviour**: The code compiles, but it fails at runtime. Which makes the type unreliable, since the point of the compiler is to catch errors at compile time, before they are caught at run time.

To better illustrate the issue, if the lib code (src/index.js) was written in TS, it would not compile. A [simplified example](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAwgewHYigDxDAvDASlAYwQCcATAHgGcRiBLJAcwBoYBDJMAPgG5hRJYcVgBthAI1YEA1uQAqnbDAAURFOhAAuGLICU2BQDcEtUr2D1UxAGaTYsqNTmLEajAoDewGMtTUtQ0QlpOU4dLSMTXgBfPgB6WJgQAAtaShgbWmFKYCsAVyQCEFpkGEpc2lQnHBdUNxVkWs1tRXcosO0HEBCYT29iKBBc4iQfTr0sDy9vRM761xAdKZiooA) would be:
```ts
function suite<T = Context>(context: T = {}): Test<T> {
  return (test) => {
    test(context)
  }
}
```

One simple, but not ideal, solution would be to [always provide a context](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAwgewHYigDxDAvDASlAYwQCcATAHgGcRiBLJAcwBoYBDJMAPgG5hRJYcVgBthAI1YEA1uQAqnbDAAURFOhAAuGLICU2BQDcEtUr2D1UxAGaTYsqNTmLEajAoDewGMtTUtQ0QlpOU4dLSMTXgBfPisAVyQCEFpkGEo42lQnHBdUNxVkPM1tMO0HEBCYT29iKBA44iQfcr0sDy9vGF8QAtcQHQ6YmOBVahgAd0yACwBRAFsIcFz1RXTMqCV3KJ1eSZBZhaXC9SVeotb271GEYSgAOmEEBjP1HZgAeneJ2lEYMVgth1PjBVEU7lYEAhFABGbgfL4AWgRCn2tEoEwQcWEpBgNh+LEo9AIsH2sFBKzmcTG-zYTSghzAMAQYgAVoQQMBtmZgaNMHjhPQGKsMll3DADCI4lAtEg4nN-sQYFFOJttvCYEiUVM0RisTj+QSibAJcIpTAdXM0YTGHxed99gA1SWwHBrUXi50yuUKpUqsUmqVaAAMSrewM1XW16PGJCklGAeymTtNGxeGAuVQ611uDyeaZAdwD9xACAAyjRBUodDtOTXgDzkGNE8mza6RRt-Z6YCGueqI1AAI4ZE1QFD2qaR2CsMQIAxQFiyUsg9jmpBWKCK0ldAS44gIOaT1eZWgiEHHDDAIA):
```ts
function suite<T = Context>(context: T): Test<T> {
  // same code as above
};
```

Another solution, in order to keep the context optional, as it is now, is [to make it `Partial`](https://www.typescriptlang.org/play?#code/FAFwngDgpgBAwgewHYigDxDAvDASlAYwQCcATAHgGcRiBLJAcwBoYBDJMAPgG5hRJYcVgBthAI1YEA1uQAqnbDAAURFOhAAuGLICU2BQDcEtUr2D1UxAGaTYsqNTmLEajAoDewGMtTUtQ0QlpOU4dLSMTXgBfPisAVyQCEFpkGEo42lQnHBdUNxVkPM0YAAVWYmSREMV3KLDtBxByMoraKvkPLxhiKBA44iQfRr0sTu9vXxAC1xAdLpiY4FVqGAB3TIALAFEAWwhwXPVFdMyoJVqdXnWQbb2DwvUlaaKRsZhlhGEoADphBAZnupLjAAPQgta0UQwMSwWpdMHvB4Yb5WBAIRQARm4oPBAFpcQobrRKGsEHFhKQYDZISxKPQCLAbrBVEUYDs4isYWxBlA7mAYAgxAArQggYB1MzLTDU4T0BjHDJZdwwAwiOJQLRIOI7GHEGBRTjnOo4mD4wkbYkwJAIVak4hSSi0+mMjbMpGYS0QcqVYR8KUQm4ANTVsBwJyVKpDmu1uv1huVquE6q0AAZ9cCEWaYESSasSA7gNcNsGk2dARhXjBPN4ER8vr9-uWQN9E+rviAEABlGhypQ6DN4gnZi0k622mWOtLOxEzFshmCehCoFBtUT8hKkKBWehQUjiy58IA):
```ts
function suite<T = Context>(context: Partial<T> = {}): Test<Partial<T>> {
  // same code as above
};
```
But this poses the issue that there will need to be undefined guards in the code (e.g. `console.log(context.value?.toString() ?? 'no value')`).

The initial solution I proposed is to have the best of both worlds. But now, while trying to write the code in TS directly, I couldn't find a way to actually represent those type declarations.

So, to summarise, I guess the main challenge is to make the context optional in the suite constructor. In order to be optional for any arbitrary `T`, it must be `Partial<T>` and default to `{}`.

Hope my explanation was more clear this time, sorry for the long reply.
